### PR TITLE
Publish support in SDK for Native Aot Application

### PR DESCRIPTION
**Pull request by LakshanF**
Add support to publish `NativeAot `applications via the SDK with 3 parts:

1. Integrate package that contain mostly SDK assets, `Microsoft.DotNet.ILCompiler`, to the SDK via bundling. This is similar to how `Microsoft.NET.ILLink.Tasks` package is leveraged.

- Add details to `Version.Details.xml, Versions.props, and BundledSdks.targets`
- Requires the `Microsoft.DotNet.ILCompiler` package to contain a `sdk.targets` in a SDK directory. This needs to be added in the runtime repo where this package is built

2. Dynamically download the package, `Runtime.**RID**.Microsoft.DotNet.ILCompiler` which contains tools, runtime and Framework libraries.
- Add *Microsoft.DotNet.ILCompiler package details to [OverrideAndCreateBundledNETCoreAppPackageVersion.cs](https://github.com/LakshanF/sdk/blob/7f8b0ce0ae7f024a94b4c1a4036de4c345f79fda/src/Layout/toolset-tasks/OverrideAndCreateBundledNETCoreAppPackageVersion.cs#L127-L134), specifically by adding a new element node and by passing `Stage0MicrosoftNETCoreAppRefPackageVersionPath` 
- Add code to download the package to [ProcessFrameworkReferences.cs](https://github.com/LakshanF/sdk/blob/7f8b0ce0ae7f024a94b4c1a4036de4c345f79fda/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs)
- Pass in publish property to [Microsoft.NET.Sdk.FrameworkReferenceResolution.targets](https://github.com/LakshanF/sdk/blob/7f8b0ce0ae7f024a94b4c1a4036de4c345f79fda/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets#L102) to invoke the `ProcessFrameworkReferences `task and get back package details, `ResolvedILCompilerPack`
- Required a [patch](https://github.com/dotnet/installer/pull/13756/files) from the installer that sets the item `KnownILCompilerPack` for the standalone SDK to work since the above change seems to only work in the SDK repo environment

3. Add a test scenario to, `GivenThatWeWantToPublishAnAotApp.cs`, to validate that PublishAot produces a native application. Linux support via running the tests in a container as well as additional test scenarios for native libraries (static and shared) have been added in subsequent PRs

Assumes that there is a `darc` channel to get the required package between runtime repo and sdk

### CODE REVIEW

File: eng/Version.Details.xml [hunk 1]
```diff
@@ -130,6 +130,11 @@
       <Sha>da3c743a606b300c245c312c8cd9b3238d110d65</Sha>
       <SourceBuild RepoName="linker" ManagedOnly="true" />
     </Dependency>
+    <Dependency Name="microsoft.dotnet.ilcompiler" Version="7.0.0-preview.4.22201.3">
```
Comment (sbomer) [inline eng/Version.Details.xml:133]: nit: 
```suggestion
    <Dependency Name="Microsoft.DotNet.ILCompiler" Version="7.0.0-preview.4.22201.3">
```

File: src/Layout/redist/targets/BundledSdks.targets [hunk 2]
```diff
@@ -5,5 +5,6 @@
     <BundledSdk Include="FSharp.NET.Sdk" Version="1.0.4-bundled-0100" />
     <BundledSdk Include="Microsoft.Docker.Sdk" Version="1.1.0" />
     <BundledSdk Include="Microsoft.NET.ILLink.Tasks" Version="$(MicrosoftNETILLinkTasksPackageVersion)" />
+    <BundledSdk Include="Microsoft.DotNet.IlCompiler" Version="$(MicrosoftNETILCompilerPackageVersion)" />
```
Comment (sbomer) [inline src/Layout/redist/targets/BundledSdks.targets:8]:
```suggestion
    <BundledSdk Include="Microsoft.DotNet.ILCompiler" Version="$(MicrosoftNETILCompilerPackageVersion)" />
```

File: src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs [hunk 3]
```diff
@@ -534,6 +550,11 @@ private bool KnownFrameworkReferenceAppliesToTargetFramework(NuGetFramework know
             }
         }
 
+        private enum AotPackageType
+        {
+            Crossgen2,
```
Comment (sbomer) [inline src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs:555]: Looks like the intent was to share the logic between crossgen2?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILCompiler.targets [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<!--
+***********************************************************************************************
+Microsoft.NET.ILCompiler.props
```
Comment (sbomer) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILCompiler.targets:3]:
```suggestion
Microsoft.NET.ILCompiler.targets
```

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILCompiler.targets [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Import Project="Sdk.props" Sdk="Microsoft.DotNet.IlCompiler" />
```
Comment (sbomer) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILCompiler.targets:14]:
```suggestion
  <Import Project="Sdk.props" Sdk="Microsoft.DotNet.ILCompiler" />
```

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunAnAotApp.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+{
+    public class GivenThatWeWantToRunAnAotApp : SdkTest
+    {
+        public GivenThatWeWantToRunAnAotApp(ITestOutputHelper log) : base(log)
+        {
+        }
+
+        [RequiresMSBuildVersionTheory("17.0.0.32901")]
+        [InlineData(LatestTfm)]        
+        public void NativeAot_only_runs_when_switch_is_enabled(string targetFramework)
```
Comment (tlakollo) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunAnAotApp.cs:37]: Given that there is a file that tests Aot apps I will include it there instead and verify that the warnings are generated 
https://github.com/dotnet/sdk/blob/main/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs

**Comment by tlakollo** [comment]
Also, we are missing something that enables the analyzer (EnableAotAnalyzer) if PublishAot is set to true

**Review by agocke — COMMENTED** [review]
I realized when looking through darc -- we already have a subscription here for the other runtime stuff. No need to set up anything new

File: src/Layout/toolset-tasks/OverrideAndCreateBundledNETCoreAppPackageVersion.cs [hunk 7]
```diff
@@ -123,6 +123,17 @@ void CheckAndReplaceAttribute(XAttribute attribute)
                 crossgen2Rids.Value += ";osx-x64";
             }
 
+            // Adding ilcompiler details package without waiting for the Stage0MicrosoftNETCoreAppRefPackageVersionPath iteration
+            XElement ilcInfo = new XElement(ns + "KnownILCompilerPack",
+                new XAttribute("Include", "Microsoft.DotNet.ILCompiler"),
+                new XAttribute("TargetFramework", "net7.0"),
+                new XAttribute("ILCompilerPackNamePattern", "runtime.**RID**.Microsoft.DotNet.ILCompiler"),
+                new XAttribute("ILCompilerPackVersion", microsoftNETCoreAppRefPackageVersion),
+                new XAttribute("ILCompilerRuntimeIdentifiers", "linux-musl-x64;linux-x64;win-x64;linux-arm;win-arm64")
```
Comment (agocke) [inline src/Layout/toolset-tasks/OverrideAndCreateBundledNETCoreAppPackageVersion.cs:132]: linux-arm64

**Review by agocke — APPROVED** [review]
LGTM aside from comment

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs [hunk 8]
```diff
@@ -102,5 +139,25 @@ static void ProduceLinkerAnalysisWarning()
...
+            using (var peReader = new PEReader(fs))
+            {
+                try
+                {
+                    var metadataReader = peReader.GetMetadataReader();
+                }catch(InvalidOperationException)
+                {
+                    returnValue = true;
+                }
+                catch(Exception){}
```
Comment (MichalStrehovsky) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs:156]: Nit: `peReader.HasMetadata` is the property to use here.

**Pull request merged**